### ISSUE TITLE
Tracking issue for migrating all remaining post-expansion feature gates to pre-expansion ones

### ISSUE THREAD

**Issue by fmease** [rust-lang/rust#154045]
> [!TIP]
> If you've arrived here because you encountered a **feature gate warning** for unstable syntax behind `#[cfg(…)]`, here's how to you can transform your code while keeping the unstable syntax `cfg`-conditional:
>
> **Before** (triggers a feature gate warning):
>
> ```rs
> #[cfg(feature = "nightly")] // or a different cfg-condition
> impl !Trait for Type {}     // or a different unstable syntax
> ```
>
> **After** (idiomatic form; doesn't trigger a warning):
>
> ```rs
> macro_rules! generate {
>     // Of course, you can use macro metavariables to avoid code duplication etc.
>     // However, keep in mind that the unstable syntax must appear in the body.
>     () => { impl !Trait for Type {} };
> }
>
> #[cfg(feature = "nightly")]
> generate!();
> ```
>
> If you'd like to see a more realistic example, head on over to https://github.com/arcnmx/packed-rs/pull/1/changes.

### Steps

1. Find the most recent relevant crater runs and manually reaudit affected crates
   * check if at least *some* crates have been fixed downstream by now 
2. Collect all features where the crater run has happened a long time ago & issue a single crater run for all of them at once
3. ~~For all the other ones that *also* don't emit a pre-expansion warning yet, add warnings immediately~~ (done)
   * ~~no further input from T-compiler/T-lang should be needed since it's already legitimized by [MCP 535].~~

### Affected Syntactic Constructs

1. item modifier `default` (feature `specialization` or `min_specialization` for fns only)
   * we potentially want to drop this feature in favor of a different approach...
   * ...so a syntax gate would make it easier to phase out the syntax later on
   * [x] issue a pre-expansion feature gate *warning* (legitimized by [MCP 535])
   * [ ] turn the warning into a hard error (needs crater, T-lang FCP)
2. trait impl modifier `!` (feature `negative_impls`)
   * likely to be stabilized in 1–3 years, so a syntax gate isn't super pressing
   * [x] issue a pre-expansion feature gate *warning* (legitimized by [MCP 535])
   * [ ] turn the warning into a hard error (needs crater, T-lang FCP)
3. expression `try { … }` (feature `try_blocks`)
   * already has a pre-expansion feature gate *warning*
   * incredibly likely to be stabilized this year...
   * ...so a syntax gate would be counterproductive / harmful (!)
   * [ ] turn the warning into a hard error (needs crater, T-lang FCP)
4. pattern `box $pat` (feature `box_patterns`)
   * already has a pre-expansion feature gate *warning*
   * incredibly likely to be replaced with explicit or implicit deref patterns...
   * ...so a syntax gate would make it easier to phase out the syntax later on
   * [ ] turn the warning into a hard error (needs crater, T-lang FCP)
5. pattern `$patpath { box $ident }`, pseudo binding mode (feature `box_patterns`)
   * incredibly likely to be replaced with explicit or implicit deref patterns...
   * ...so a syntax gate would make it easier to phase out the syntax later on
   * [x] issue a pre-expansion feature gate *warning* (legitimized by [MCP 535])
   * [ ] turn the warning into a hard error (needs crater, T-lang FCP)
6. item `trait $ident $genericparams = $bounds;` (feature `trait_alias`)
   * already has a pre-expansion feature gate *warning*
   * [ ] turn the warning into a hard error (needs crater, T-lang FCP)
7. trait item modifier `auto` (feature `auto_traits`)
   * already has a pre-expansion feature gate *warning*
   * [ ] turn the warning into a hard error (needs crater, T-lang FCP)
8. (not fitting 100%) syntax `impl $traitref for .. {}` (early version of a now removed predecessor of feature `auto_traits`)
   * most recent crater run: PR RUST-121072
   * [ ] issue a pre-expansion feature gate *warning* (legitimized by [MCP 535], maybe?)
   * [ ] reevaluate RUST-121072's crater report & send downstream patches
   * [ ] turn the warning into a hard error (needs T-lang FCP, maybe another crater run)
9. item `macro $ident$macroparams { $ttstar }` (feature `decl_macro`)
   * already has a pre-expansion feature gate *warning*
   * [ ] turn the warning into a hard error (needs crater, T-lang FCP)
10. (not fitting 100%) predicates `$ty = $ty` and `$ty == $ty` (no corresp. feature was ever added (cc RUST-22074, RUST-87471))
    * [ ] straight up hard error (crater: 0 regressions, needs T-lang FCP): PR RUST-153513

### Pre-History (Incomplete)

* https://github.com/rust-lang/rust/pull/63545
* https://github.com/rust-lang/rust/pull/65742
* https://github.com/rust-lang/rust/issues/65860
* https://github.com/rust-lang/rust/pull/66004
* https://github.com/rust-lang/compiler-team/issues/535
* https://github.com/rust-lang/rust/pull/99935
* https://github.com/rust-lang/rust/pull/116393
* https://github.com/rust-lang/rust/pull/121072

### Implementation History

* https://github.com/rust-lang/rust/pull/153513
* https://github.com/rust-lang/rust/pull/154475
* https://github.com/rust-lang/rust/pull/154527

#### Downstream Patches

* https://github.com/arcnmx/packed-rs/pull/1

[MCP 535]: https://github.com/rust-lang/compiler-team/issues/535

**Comment by fmease** [comment]
Thanks for reaching out!

I admit that the warning doesn't give any real guidance for how to address the issue (that affects all unstable syntax warnings and should be rectified). We have https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unstable-syntax-pre-expansion (§ *Explanation*) which offers a vague solution sketch but it's not linked from anywhere, we should definitely improve upon that.

**The solution is to hide the `cfg`-conditional unstable syntax behind a macro invocation** since neither the macro input nor the macro body is parsed/expanded pre expansion obviously.

In your case that could look something like:

```rs
macro_rules! generate {
    () => { impl !Send for NotSend {} }
}

#[cfg(feature = "nightly_tests")]
generate!();
```

Or like this:

```rs
macro_rules! generate { ($( $tt:tt )*) => { $( $tt )* } }

#[cfg(feature = "nightly_tests")]
generate! { impl !Send for NotSend {}  }
```

See also downstream PR https://github.com/arcnmx/packed-rs/pull/1/changes in which I perform a similar transformation.

You may ask why we force you to write it in this convoluted way going forward, there's a good reason: If we ever changed the syntax of negative impls (or any feature really) or removed it entirely (which we should arguably be allowed to do on grounds of it being *unstable*), your code would break even if it previously compiled on stable (thanks to the `cfg`) which would violate our stability guarantees. Of course, for now it's only a warning as a preparation for a hard error (in a few years unless the feature gets stabilized before that).

For context, *every* new unstable syntactic feature since probably 2019 will raise an error even behind `cfg` precisely for that reason. Only 9 unstable features still have a feature gate that can get suppressed by a `cfg` since they were introduced before the new policy. They are de facto stable despite officially being unstable which makes it incredibly hard to "destabilize" them again. It'll probably take half a decade or longer, and it's already been 7 years.

Putting unstable syntax behind `#[cfg(feature = "nightly")]` or similar is a common pattern by the way it just needs to be done slightly differently now.

**Comment by petrochenkov** [comment]
`stmt_expr_attributes` is also supposed to be a pre-expansion feature gate, but it's not.
